src/ray/object_manager/ownership_based_object_directory.cc

```diff
@@ -0,0 +1,329 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/object_manager/ownership_based_object_directory.h"
+
+namespace ray {
+
+OwnershipBasedObjectDirectory::OwnershipBasedObjectDirectory(
+    boost::asio::io_service &io_service, std::shared_ptr<gcs::GcsClient> &gcs_client)
+    : io_service_(io_service),
+      gcs_client_(gcs_client),
+      client_call_manager_(io_service_) {}
+
+namespace {
+
+using ray::rpc::GcsChangeMode;
+using ray::rpc::GcsNodeInfo;
+using ray::rpc::ObjectTableData;
+
+/// Process a notification of the object table entries and store the result in
+/// node_ids. This assumes that node_ids already contains the result of the
+/// object table entries up to but not including this notification.
```

Comment:
Is it possible to inherit from the base object directory so we don't need to copy paste these methods?